app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmReceiver.kt

```diff
@@ -0,0 +1,211 @@
+package nerd.tuxmobil.fahrplan.congress.alarms
+
+import nerd.tuxmobil.fahrplan.congress.utils.PendingIntentCompat.FLAG_IMMUTABLE
+import nerd.tuxmobil.fahrplan.congress.contract.BundleKeys
+import nerd.tuxmobil.fahrplan.congress.schedule.MainActivity.Companion.createLaunchIntent
+import nerd.tuxmobil.fahrplan.congress.exceptions.BuilderException
+import nerd.tuxmobil.fahrplan.congress.autoupdate.UpdateService.Companion.start
+import nerd.tuxmobil.fahrplan.congress.repositories.AppRepository
+import nerd.tuxmobil.fahrplan.congress.notifications.NotificationHelper
+
+import info.metadude.android.eventfahrplan.commons.logging.Logging
+
+import android.net.Uri
+import android.app.PendingIntent
+import android.content.Intent
+import android.content.Context
+import android.content.BroadcastReceiver
+import androidx.annotation.VisibleForTesting
+
+class AlarmReceiver : BroadcastReceiver() {
+
+    companion object {
+        const val ALARM_UPDATE = "nerd.tuxmobil.fahrplan.congress.ALARM_UPDATE"
+        const val ALARM_SESSION = "nerd.tuxmobil.fahrplan.congress.ALARM_SESSION"
+        const val ALARM_DELETE = "de.machtnix.fahrplan.ALARM"
+        const val ALARM_DISMISSED = "nerd.tuxmobil.fahrplan.congress.ALARM_DISMISSED"
+        private const val LOG_TAG = "AlarmReceiver"
+        private const val BUNDLE_KEY_NOTIFICATION_ID = "BUNDLE_KEY_NOTIFICATION_ID"
+        private const val INVALID_NOTIFICATION_ID = -1
+        private const val DEFAULT_REQUEST_CODE = 0
+
+        /**
+         * Returns a unique [Intent] to delete the data associated with the
+         * given session alarm `notificationId`.
+         * The `Intent` is composed with a fake data URI to comply with the uniqueness rules
+         * defined by [Intent.filterEquals].
+         */
+        private fun createDeleteNotificationIntent(context: Context, notificationId: Int) =
+            Intent(context, AlarmReceiver::class.java).apply {
+                action = ALARM_DISMISSED
+                putExtra(BUNDLE_KEY_NOTIFICATION_ID, notificationId)
+                data = Uri.parse("fake://$notificationId")
+            }
+
+        private fun getAddAlarmIntent(
+            context: Context,
+            sessionId: String,
+            day: Int,
+            title: String,
+            startTime: Long
+        ) = getAlarmIntent(context, sessionId, day, title, startTime, ALARM_SESSION)
+
+        private fun getDeleteAlarmIntent(
+            context: Context,
+            sessionId: String,
+            day: Int,
+            title: String,
+            startTime: Long
+        ) = getAlarmIntent(context, sessionId, day, title, startTime, ALARM_DELETE)
+
+        private fun getAlarmIntent(
```

Comment:
I don't think we need all these helper functions. Please simplify when you move them into the factory.

Reply:
I think you can **move** and **inline** the helper functions within the factory class.

Reply:
I saw that, but there are no helper functions that could be moved. The setters for `context`, `sessionId`, `title`, `day` and `startTime` are obsolete because the properties are final, and the kind of intent (add/delete) is defined via a the `isAddAlarmIntent` parameter of the `getIntent`-Method

Reply:
Works for me: 

``` kotlin
internal class AlarmIntentFactory(
    val context: Context,
    val sessionId: String,
    val title: String,
    val day: Int,
    val startTime: Long,
) {

    fun getIntent(isAddAlarmIntent: Boolean) = Intent(context, AlarmReceiver::class.java)
        .withExtras(
            BundleKeys.ALARM_SESSION_ID to sessionId,
            BundleKeys.ALARM_DAY to day,
            BundleKeys.ALARM_TITLE to title,
            BundleKeys.ALARM_START_TIME to startTime
        ).apply {
            action = if (isAddAlarmIntent) ALARM_SESSION else ALARM_DELETE
            data = "alarm://$sessionId".toUri()
        }
```

Reply:
The constants can then be marked as `@VisibleForTesting`.